src/app/zap-templates/templates/app/cluster-objects-src.zapt

```diff
@@ -58,6 +59,29 @@ CHIP_ERROR DecodableType::Decode(TLV::TLVReader &reader) {
         {
             {{#zcl_command_arguments}}
             case to_underlying(Fields::k{{asUpperCamelCase label}}):
+                {{#if_is_enum type}}
+                {{#unless isArray}}
+                {{#if (isWeaklyTypedEnum (asUpperCamelCase type))}}
+// Need to convert consumers to using the new enum classes, so we
+// don't just have casts all over.
+#ifdef CHIP_USE_ENUM_CLASS_FOR_IM_ENUM
+                {{/if}}
+                {{#if isNullable}}
+                VerifyOrReturnError(!{{asLowerCamelCase label}}.IsNull(), CHIP_NO_ERROR);
+                ReturnErrorOnFailure(Check{{asUpperCamelCase type}}({{asLowerCamelCase label}}.Value()));
+                {{else}}
+                {{#if isOptional}}
+                VerifyOrReturnError({{asLowerCamelCase label}}.HasValue(), CHIP_NO_ERROR);
+                ReturnErrorOnFailure(Check{{asUpperCamelCase type}}({{asLowerCamelCase label}}.Value()));
+                {{else}}
+                ReturnErrorOnFailure(Check{{asUpperCamelCase type}}({{asLowerCamelCase label}}));
+                {{/if}}
+                {{/if}}
+                {{#if (isWeaklyTypedEnum (asUpperCamelCase type))}}
+#endif
+                {{/if}}
+                {{/unless}}
+                {{/if_is_enum}}
                 ReturnErrorOnFailure(DataModel::Decode(reader, {{asLowerCamelCase label}}));
```

Comment:
So I think what we should be doing is that the Check functions should be called from _inside `Decode` for enums_.  That will make it work right for enums as command fields, enums as struct fields, lists of enums, enum-typed attributes, nullable enums, optional enums, etc, etc, etc.

Reply:
updated, thanks